### PR TITLE
[IMP] chart: Add combo chart

### PR DESCRIPTION
**Pull request by anhe-odoo**
## Task Description

This task aims to add the combo chart from chartJS. As it's almost the same as the bar chart, with only another parameters telling the type of each dataset, this type of chart has been implemented in a way to keep as much code as possible common to both charts.
This PR add also an option to use the right axis for the line part of the datasets.

## Related Task

- Task: [3648579](https://www.odoo.com/web#id=3648579&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/helpers/figures/charts/chart_common_bar_combo.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        ? this.getters.getRangeString(this.labelRange, this.sheetId)
+        : undefined,
+    };
+  }
+
+  copyForSheetId(sheetId: UID): ComboBarChart {
+    const dataSets = copyDataSetsWithNewSheetId(this.sheetId, sheetId, this.dataSets);
+    const labelRange = copyLabelRangeWithNewSheetId(this.sheetId, sheetId, this.labelRange);
+    const definition = this.getDefinitionWithSpecificDataSets(dataSets, labelRange, sheetId);
+    return new ComboBarChart(this.type, definition, sheetId, this.getters);
```
Comment (hokolomopo) [inline src/helpers/figures/charts/chart_common_bar_combo.ts:115]: That won't work. That means that when we duplicate a combo chart, the ComboChart class instance becomes a `ComboBarChart`. So we never call the constructor, lose `useBothYAxis` and call the wrong `getDefinition`.

`ComboBarChart` should be an abstract class

File: src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  get stackedLabel(): string {
+    return _t("Stacked combochart");
+  }
+
+  get shouldUseRightAxis(): string {
+    return _t("Use right axis for line series");
+  }
+
+  onUpdateStacked(stacked: boolean) {
+    this.props.updateChart(this.props.figureId, {
```
Comment (hokolomopo) [inline src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts:16]: omega nitpick: this could be prettified to one line

**Review by hokolomopo — COMMENTED** [review]
Really cool chart :smile:

File: src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts (hunk 2 above)
Comment (anhe-odoo) [reply src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts:16]: In fact ti should be removed, as, as you said, stacked combo chart is a non-sense :)

File: demo/main.js [hunk 3]
```diff
@@ -189,6 +189,12 @@ class Demo extends Component {
             inputFiles[images[i]] = { imageSrc };
           }
           await this.initiateConnection(inputFiles);
+          stores.resetStores();
+          stores.inject(NotificationStore, {
+            notifyUser: this.notifyUser,
+            raiseError: this.raiseError,
+            askConfirmation: this.askConfirmation,
+          });
```
Comment (pro-odoo) [inline demo/main.js:197]: Why is it necessary ? It's seems to not be related to combo chart and it's rather a fix ? I would put it in a dedicated commit

File: src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { _t } from "../../../../translation";
+import { LineBarPieConfigPanel } from "../line_bar_pie_panel/config_panel";
+
+export class ComboConfigPanel extends LineBarPieConfigPanel {
```
Comment (pro-odoo) [inline src/components/side_panel/chart/combo_chart/combo_chart_config_panel.ts:4]:
```suggestion
export class ComboChartConfigPanel extends LineBarPieConfigPanel {
```
To be consistent with `ComboChartDesignPanel`

File: src/helpers/figures/charts/combo_chart.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  getChartDatasetFormat,
+  getChartDatasetValues,
+  getChartLabelValues,
+  getDefaultChartJsRuntime,
+} from "./chart_ui_common";
+
+export class ComboChart extends AbstractChart {
+  readonly useBothYAxis?: boolean;
+  readonly dataSets: DataSet[];
+  readonly labelRange?: Range | undefined;
```
Comment (pro-odoo) [inline src/helpers/figures/charts/combo_chart.ts:53]:
```suggestion
  readonly labelRange?: Range;
```

File: src/helpers/figures/charts/combo_chart.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        ? this.getters.getRangeString(labelRange, targetSheetId || this.sheetId)
+        : undefined,
+      title: this.title,
+      aggregated: this.aggregated,
+      useBothYAxis: this.useBothYAxis,
+    };
+  }
+
+  getDefinitionForExcel(): ExcelChartDefinition | undefined {
+    if (this.aggregated) return undefined;
```
Comment (pro-odoo) [inline src/helpers/figures/charts/combo_chart.ts:133]: Worth a little comment

File: src/types/chart/bar_chart.ts [hunk 7]
```diff
@@ -1,21 +1,6 @@
...
-  readonly labelRange?: string;
-  readonly title: string;
-  readonly background?: Color;
-  readonly verticalAxisPosition: VerticalAxisPosition;
-  readonly legendPosition: LegendPosition;
   readonly stacked: boolean;
-  readonly aggregated?: boolean;
 }
-
-export type BarChartRuntime = {
```
Comment (pro-odoo) [inline src/types/chart/bar_chart.ts:18]: IMO it's better to keep `BarChartRuntime` and `ComboChartRuntime` even if it's the same type, I think it's clearer. WDYT ?

**Review by pro-odoo — COMMENTED** [review]
👍

File: demo/main.js (hunk 3 above)
Comment (anhe-odoo) [reply demo/main.js:197]: This part of the code is from a PR of Adrien, so I just added it to my code to use it for testing the export part

File: src/types/chart/bar_chart.ts (hunk 7 above)
Comment (anhe-odoo) [reply src/types/chart/bar_chart.ts:18]: I didn't have any preference, I just did it that way to avoid duplicating stuff, but you may be right :)

**Comment by anhe-odoo** [comment]
@pro-odoo : I took your comment into account and added the tests for the xlsx import export part too :) You can have a second look if you want/have some time

**Review by LucasLefevre — COMMENTED** [review]
Looking good
Could add at least a few side panel tests?

**Review by LucasLefevre — APPROVED** [review]
robodoo r+
